sgx_urts/src/net.rs

```diff
@@ -15,34 +15,72 @@
 // specific language governing permissions and limitations
 // under the License..
 
+use libc::{self, addrinfo, c_char, c_int, size_t};
+use std::ffi::CStr;
 use std::io::Error;
-use libc::{self, c_int, c_char, addrinfo};
+use std::slice;
+
+use core::ptr;
 
 #[no_mangle]
 pub extern "C" fn u_getaddrinfo_ocall(
     error: *mut c_int,
     node: *const c_char,
     service: *const c_char,
-    hints: *const addrinfo,
-    res: *mut *mut addrinfo,
+    hints: *mut addrinfo,
+    entry_size: size_t,
+    buf: *mut u8,
+    bufsz: size_t,
+    out_cnt: *mut size_t,
 ) -> c_int {
     let mut errno = 0;
-    let ret = unsafe { libc::getaddrinfo(node, service, hints, res) };
+    let mut res: *mut addrinfo = ptr::null_mut();
+
+    let ret = unsafe { libc::getaddrinfo(node, service, hints, &mut res) };
+
     if ret == libc::EAI_SYSTEM {
         errno = Error::last_os_error().raw_os_error().unwrap_or(0);
     }
     if !error.is_null() {
-        unsafe { *error = errno; }
+        unsafe {
+            *error = errno;
+        }
     }
-    ret
-}
 
-#[no_mangle]
-pub extern "C" fn u_freeaddrinfo_ocall(res: *mut addrinfo ) {
-    unsafe { libc::freeaddrinfo(res) }
+    if ret != 0 {
+        return ret;
+    }
+
+    unsafe {
+        let mut i = 0;
+        let mut cur_ptr = res;
+        while cur_ptr != ptr::null_mut() && (i + entry_size) < bufsz {
+            let cur: &addrinfo = &*cur_ptr;
+            let len = cur.ai_addrlen as usize;
+            if len > entry_size {
+                return 1;
```

Comment:
1 = `EAI_ADDRFAMILY`. recommend return `EAI_SYSTEM`

Reply:
> 1 = `EAI_ADDRFAMILY`. recommend return `EAI_SYSTEM`

I made the calling convention of `u_getaddrinfo_ocall` different from `libc::getaddrinfo` here. More specifically, 
 `ret == 0` means success; if ret is negative, check the errno for the error information; positive ret value is used for errors introduced by our shim layer logic. In this case, 1 means the addrinfo entry size expecting by the caller is too small to hold the actual entry information. 

The reason I am doing this is because I cannot find any existing error number describing this kind error. To make it more understandable, I think we can define some meaningful constants to represent shim layer error like this and add documents for the API conventions.

Reply:
👌👌